### PR TITLE
kv-cache: optional per-channel K-cache mean-centering for Q4_0

### PR DESCRIPTION
**Pull request by bri-prism**
## Summary

Adds an opt-in K-cache mean-centering feature: a fixed, precomputed per-(kv-head, channel) bias is subtracted from the K vector right before it is quantized into the cache. This is currently scoped strictly to `GGML_TYPE_Q4_0` K caches.

## The idea

`GGML_TYPE_Q4_0` is a symmetric (zero-point-free) block quantizer. If a given (kv-head, channel) position's real K activations have a nonzero mean across tokens, symmetric quantization wastes some of its dynamic range encoding that constant bias, which increases quantization error for that channel. Centering the residual around zero before quantization reduces that error for channels with a real, consistent activation bias.

## Why this is safe

For a fixed query row `q` attending over cached keys `k_0 ... k_n`, if every cached key is centered by the same bias `k_bar` before being quantized and stored:

```
q . k_i = q . (k_i - k_bar) + q . k_bar = q . k_i_stored + q . k_bar
```

The `q . k_bar` term doesn't depend on the key's position, so it's added identically to every logit in that query's row. Softmax is invariant to a constant additive shift applied to every logit in the same row, so nothing else in attention needs to change. This also means it's a zero decode-time-cost win: one subtract at the point of cache write.

`tests/test-kv-mean-center.cpp` checks this directly: against an unquantized F32 K cache (to avoid confounding the comparison with real quantization error), a synthetic nonzero bias applied through the exact same code path leaves output logits matching the uncentered baseline to fp32 rounding (nmse ~5e-10 in practice).

## What's included

- **Core mechanism** (`src/llama-kv-cache.{h,cpp}`, `src/llama-context.cpp`, `include/llama.h`): a new `path_kv_mean_center` field on `llama_context_params`, applied in `llama_kv_cache::cpy_k()` right before the K value is quantized and stored. Gated strictly to `k->type == GGML_TYPE_Q4_0`; if the flag is set but the cache type doesn't match, context creation hard-fails with a clear error, matching this codebase's existing convention for other cache-type-gated mismatches (e.g. "V cache quantization requires flash_attn").
- **CLI flag** (`common/`): `--kv-mean-center <path>`, requires `-ctk q4_0`.
- **Bias file format**: a small GGUF file with one F32 tensor per layer (`kv_bar.blk.<il>.k`), shaped to broadcast against the K tensor's natural `[n_embd_head, n_head]` layout. Writer lives in `common/kv-mean-center.*` so it's shared between the calibration tool and the tests.
- **Calibration tool** (`tools/kv-mean-center/`): follows the `tools/imatrix` convention. Runs a text calibration corpus through the model and captures the K tensor right before it's written into the cache, via a one-line addition of a `cb()` graph-build tag (`"k_cache_in"`, in `llm_graph_context::build_attn()`) read through the same backend-scheduler eval-callback mechanism `llama-imatrix` uses.
- **Tests** (`tests/test-kv-mean-center.cpp`): regression safety (centering disabled is a byte-for-byte no-op, including for a Q4_0 cache with no bias file loaded), the `--kv-mean-center`/Q4_0 gate itself, and the softmax-invariance math check described above.
- **Docs** (`docs/kv-mean-center.md`, `tools/kv-mean-center/README.md`).

## Scope / follow-ups

- Only `GGML_TYPE_Q4_0` is supported; other quant types are rejected. This isn't an arbitrary narrowing: Q4_0 is the only commonly-used int4 K-cache type that's symmetric (zero-point-free), which is the precondition for this technique to help at all. The other common int4 option, Q4_1, is asymmetric (it already stores a per-block min/bias alongside its scale), so it already absorbs a static per-channel mean the same way MLX's affine quantization does — centering would be a measurable no-op there. Generalizing further (e.g. to Q5_0, another symmetric type) is future work, but there's no other *int4* type this would meaningfully apply to.
- Only the plain (non-recurrent, non-hybrid, non-MLA/DSA) KV cache is supported.
- The `k_cache_in` calibration hook currently only covers the standard dense/GQA attention path, which is the large majority of architectures but not all of them.
- If this fork's optional Hadamard K/Q rotation is also active, the bias is calibrated pre-rotation but applied post-rotation; this stays exactly safe (the invariance argument is basis-independent) but is a less accurate calibration in that specific configuration. Noted in the docs as a follow-up.
- Quantifying the real quantization-fidelity improvement on a production-scale model (e.g. a logit-KLD comparison against an uncentered Q4_0 baseline) is a natural follow-up; this PR doesn't include a measured number for any specific trained model.

## Test plan

- [x] `cmake --build build --target llama` (core lib) builds clean
- [x] Full project build (`llama-cli`, `llama-server`, all tools/tests) builds clean
- [x] `ctest -R test-kv-mean-center` passes (regression safety, gate, invariance)
- [x] `ctest -R test-llama-archs` (existing test, unaffected) passes
- [x] Confirmed the 3 unrelated pre-existing test failures in this environment (`test-tokenizers-ggml-vocabs`, `test-quantize-fns`, `test-quant-type-selection`) reproduce identically on the unmodified base branch, i.e. are not caused by this change

### CODE REVIEW

**Comment by bri-prism** [comment]
Addressed the 6 Copilot review comments in 255147556:

- tools/kv-mean-center/kv-mean-center.cpp (comments on the F32 assert and the raw-byte reinterpret): the K tensor captured at the "k_cache_in" tag can be F16 or BF16 depending on backend/compute settings, not only F32. The collector now accepts F32/F16/BF16 and converts to F32 with ggml_fp16_to_fp32_row/ggml_bf16_to_fp32_row before accumulating the mean, instead of asserting F32-only and reinterpreting raw bytes as float*.
- tools/kv-mean-center/README.md (the --chunks gap): wired --chunks into the LLAMA_EXAMPLE_KV_MEAN_CENTER example set in common/arg.cpp, so the flag the README documents is actually available to this tool rather than being filtered out by the shared arg parser.
- docs/kv-mean-center.md (the two non-ASCII characters): replaced the U+2248 ("approx equal") and U+00B7 (middle dot) characters with ASCII equivalents (~= and .), matching this project's ASCII-only docs convention.

On the CI failures: the three failing jobs (ubuntu x64, ubuntu arm64, windows) are pre-existing and unrelated to this PR.

- ubuntu x64 and arm64 fail in test-quantize-fns on Q2_0 ("q2_0 absolute quantization error: FAILED", "q2_0 dot product error: FAILED"). This PR only touches GGML_TYPE_Q4_0 code paths and never references Q2_0 anywhere in its diff. I re-verified this directly: checking out the actual pre-PR base commit (0ad1dab7b, the parent of this PR's first commit) and building test-quantize-fns there reproduces the identical failure with the identical error values (0.008678 / 0.141111), with none of this PR's changes present. Rebuilding at the current PR tip reproduces the exact same numbers again, confirming this PR has zero effect on that test's outcome either way.
- windows fails in unit/test_kv_keep_only_active.py::test_clear_and_restore, a server-side idle-slot-cache-clearing integration test with no relationship to K-cache mean-centering.

Not attempting to fix either of these here since they're outside this PR's scope.

**Review by khosravipasha — APPROVED** [review]
Took a pass, LGTM

some questions
1. for our fork this makes sense, do we want to eventually upstream? in that case might need a bunch of changes
2. for the mean centering needs calibration data? `tools/kv-mean-center`, doest it matter much? Mostly asking for next qestion;
3. what's plan to integrate into our bonsai-demo repo? Need clear examples and how to use there eventuallly.

also after merngin need to cut a build-release from actions eventually, so the prebuild version of demos also work out of the box.
